### ISSUE TITLE
Dependency order problem, rebuilds due to incorrect implicit dependencies

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#2811]
This issue was originally created at: 2012-01-10 10:19:23.
This issue was reported by: `dirkbaechle`.

> dirkbaechle said at 2012-01-10 10:19:23

Hi there,
 
as discussed at some length in http://scons.tigris.org/ds/viewMessage.do?dsForumId=1268&dsMessageId=2855320, SCons starts rebuilds on a second call, when obviously nothing has changed.
> 
>I started to debug this problem, with gem5 ( www.gem5.org ) as example, and it is connected to the way we scan (don't scan) for implicit dependencies. I am able to reproduce the "2nd build" on my Ubuntu 11.04 box, by repeatedly calling:
```console
rm -f build/sconsign.dblite
scons build/ALPHA_SE/python/m5/internal/param_NetworkTest.i
scons build/ALPHA_SE/python/m5/internal/param_NetworkTest.i
```

Note that on an initial build, or after the "rm -f sconsign.dblite", no signature infos are available at all.
 
So, when the Taskmaster tries to find the next "ready" node it uses `get_implicit_deps()` to find all children for the node in question. But no children are returned, because `current_check()` (`SCons/Scanner/__init__.py:318`) checks whether the node is "up_to_date". No signature info means that the node has changed, and then `__call__()` (`SCons/Scanner/__init__.py:210`) returns the emtpy list:
```py
        if self.scan_check and not self.scan_check(node, env):
            return []
```
 
Stack trace for this is:
 
```
MainThread - pid30609_seq1
    changed [__init__.py:1066]
    is_up_to_date [FS.py:3028]
    current_check [__init__.py:318]
    __call__ [__init__.py:210]
    get_found_includes [FS.py:2681]
    get_implicit_deps [__init__.py:555]
    scan [Executor.py:482]
    scan_sources [Executor.py:451]
    scan [__init__.py:648]
    children [__init__.py:999]
    get_all_children [Executor.py:261]
    _find_next_ready_node [Taskmaster.py:786]
    next_task [Taskmaster.py:918]
    start [Job.py:193]
    run [Job.py:111]
    _build_targets [Main.py:1264]
    _main [Main.py:1070]
    _exec_main [Main.py:1306]
    main [Main.py:1342]
    <module> [Main.py:1402]
    run [pydevd.py:916]
    <module> [pydevd.py:1145] 
```
Like this, only "/usr/bin/swig" is found as implicit dependency during the first run of the while-loop in get_implicit_deps() (SCons/Node/__init__.py:552).
 
When calling "scons" the second time, signature infos exist...so the SWIG scanner is actually called on the file param_NetworkTest.i and adds several includes as implicit deps *before* "/usr/bin/swig".
 
This seems to cause the change of the filename order for the dependencies, and triggers the second (probably unnecessary) build.
 
Probably the best solution for this, is to allow the scanning of source files for implicit dependencies, in the case where no sig info was computed but the file itself exists.

I will try to come up with a concrete patch and testcase for this asap.
 
Regards,
 
Dirk

>batmanaod said at 2012-06-08 08:34:22

It appears that the people working on mongoDB have found a way to deal with this error; the initial recognition of the error is here (https://jira.mongodb.org/browse/SERVER-5254) and the fix is here (https://github.com/mongodb/mongo/commit/5455f1d95c232c92647e3e0cfda4f52b28a08ef7) . Unfortunately, it appears to me that their solution is only viable because their libdeps plugin completely changes the way static libraries are made, including how dependencies are found; I haven't yet figured out how a similar change could be made to the central SCons code (though I'm not very familiar with the internal workings of SCons, so I may just not know how/where to make the change).

>dirkbaechle said at 2014-05-18 03:13:07
reassigning issue

Votes for this issue: 1.

**Comment by bdbaddog** [comment]
Looks like I have a fix for this issue.
Still working on creating a simple reproducer as currently the reproducer is the MongoDB build. (A bit too big for our regression suite).

The problem is as folllows.
Object file depends on generated source.
Generated source isn't generated when first scan of object.
Object get's only compiler as implicit target.

In this builds case the object builder is also spitting out a split-dwarf .dwo file (side file with debug info).  
When the source file is finally rebuild the pending_parents for the .cpp doesn't list the .dwo.
The build completes and sconsign is written.
The second build sets the .dwo files implicit list in proper order and then rebuilds because the order has changed..

Look for a fix in next post 3.0.5 release.

**Comment by dmoody256** [comment]
Could this be related to https://github.com/SCons/scons/issues/2908 also? Sounds similar.

**Comment by bdbaddog** [comment]
> Could this be related to #2908 also? Sounds similar.

Could be.  I have an associated enhancement to --debug=explain which can help highlight if this is the issue. I'll try to get a PR for that today and perhaps you can try and post?

**Comment by bdbaddog** [comment]
Nope this is different than #2908

**Comment by matutter** [comment]
+1

I'm experiencing the same problem when using the SCons' "protoc" Builder. Any source-files with implicit dependencies of the generated `protoc` output are _spuriously_ rebuilt and `--debug-explain` displays `scons: rebuilding "x" because the dependency order changed:`.

**Comment by bdbaddog** [comment]
> +1
> 
> I'm experiencing the same problem when using the SCons' "protoc" Builder. Any source-files with implicit dependencies of the generated `protoc` output are _spuriously_ rebuilt and `--debug-explain` displays `scons: rebuilding "x" because the dependency order changed:`.

Can you try the code in the #3340 to see if it resolves your issue?

**Comment by matutter** [comment]
I might be getting these rebuilds for some legitimate reason other than, or in addition to, this one because this didn't seem to resolve my issue. 

What did solve my issue was sorting deps by the node's `path`, almost exactly what mongodb did.

I'm going to try and see if I can reproduce the problem in less complicated project and see if #3340 can fix it.

**Comment by bdbaddog** [comment]
Mat,

So if you're sorting the deps, your discarding the ability to check the order and thus part of scons' reasonable dependency change checking.  If the order is wrong something is wrong.  I suggest you also pull the enhanced explain PR I pushed but haven't merged yet and run scons `--debug=explain` and see which part of the dependency list order has changed.

**Comment by bdbaddog** [comment]
@matutter  - Can you bring your issue the users mailing list, hopefully with small reproducer?

**Comment by bdbaddog** [comment]
fixed with #3340
